CoreFoundation/Locale.subproj/CFCalendar.c

```diff
@@ -440,669 +958,2000 @@ static UCalendarDateFields __CFCalendarGetICUFieldCodeFromChar(char ch) {
+            nth_weekday = (10 < nth_weekday) ? nth_weekday - 10 : 0; // low-ball estimate
+            do {
+                nth_weekday++;
+                status = U_ZERO_ERROR;
+                __cficu_ucal_clear(calendar->_cal);
+                __cficu_ucal_setMillis(calendar->_cal, start_udate, &status);
+                test_udate = __CFCalendarAdd(calendar, UCAL_WEEK_OF_YEAR, nth_weekday, 0, &status);
+            } while (test_udate <= at_udate);
+            return nth_weekday;
+        }
+        case kCFCalendarUnitDay: {
+            CFAbsoluteTime start = 0.0;
+            Boolean b = CFCalendarGetTimeRangeOfUnit(calendar, kCFCalendarUnitEra, at, &start, NULL);
+            // must do this to make sure things are set up for recursive calls to __CFCalendarGetOrdinalityOfUnit3
+            UErrorCode status = U_ZERO_ERROR;
+            __cficu_ucal_clear(calendar->_cal);
+            UDate udate = (floor(at) + kCFAbsoluteTimeIntervalSince1970) * 1000.0;
+            __cficu_ucal_setMillis(calendar->_cal, udate, &status);
+            if (!b) return kCFNotFound;
+            CFIndex day = (CFIndex)floor((at - start) / 86400.0) + 1;
+            return day;
+        }
+        case kCFCalendarUnitHour: {
+            UErrorCode status = U_ZERO_ERROR;
+            CFIndex day = __CFCalendarGetOrdinalityOfUnit3(calendar, kCFCalendarUnitDay, kCFCalendarUnitEra, at);
+            if (kCFNotFound == day) return kCFNotFound;
+            if ((LONG_MAX - 24) / 24 < (day - 1)) return kCFNotFound;
+            CFIndex hour = (day - 1) * 24 + __cficu_ucal_get(calendar->_cal, UCAL_HOUR_OF_DAY, &status) + 1;
+            return hour;
+        }
+        case kCFCalendarUnitMinute: {
+            UErrorCode status = U_ZERO_ERROR;
+            CFIndex hour = __CFCalendarGetOrdinalityOfUnit3(calendar, kCFCalendarUnitHour, kCFCalendarUnitEra, at);
+            if (kCFNotFound == hour) return kCFNotFound;
+            if ((LONG_MAX - 60) / 60 < (hour - 1)) return kCFNotFound;
+            CFIndex minute = (hour - 1) * 60 + __cficu_ucal_get(calendar->_cal, UCAL_MINUTE, &status) + 1;
+            return minute;
+        }
+        case kCFCalendarUnitSecond: {
+            UErrorCode status = U_ZERO_ERROR;
+            CFIndex minute = __CFCalendarGetOrdinalityOfUnit3(calendar, kCFCalendarUnitMinute, kCFCalendarUnitEra, at);
+            if (kCFNotFound == minute) return kCFNotFound;
+            if ((LONG_MAX - 60) / 60 < (minute - 1)) return kCFNotFound;
+            CFIndex second = (minute - 1) * 60 + __cficu_ucal_get(calendar->_cal, UCAL_SECOND, &status) + 1;
+            return second;
+        }
+        }
+#pragma GCC diagnostic pop // See 10693376
+         break;
+    case kCFCalendarUnitYear:
+#pragma GCC diagnostic push // See 10693376
+#pragma GCC diagnostic ignored "-Wswitch-enum"
+        switch (smallerUnit) {
+        case kCFCalendarUnitQuarter: {
+            UErrorCode status = U_ZERO_ERROR;
+    
```

Comment:
I need to investigate this, but it looks like I may be able to just drop this check.

Reply:
(I’m willing to pursue this on this patch because we want to guarantee availability of all these API on all systems that scf can target, which _is_ a correctness issue on the patch.)

Reply:
Yeah, we should ensure that CF works everywhere.